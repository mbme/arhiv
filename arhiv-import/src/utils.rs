use anyhow::{bail, Context, Result};
use async_trait::async_trait;
use dialoguer::{theme::ColorfulTheme, Confirm};

use arhiv_core::{entities::Document, Arhiv};
use rs_utils::{log, run_js_script};

#[async_trait]
pub trait Importer {
    fn get_name(&self) -> &str;

    fn can_import(&self, url: &str) -> bool;

    async fn import(&self, url: &str, arhiv: &Arhiv, confirm: bool) -> Result<Document>;
}

fn scrape(args: Vec<&str>) -> Result<String> {
    let script = include_str!("../dist/bundle.js");

    run_js_script(script, args)
}

pub fn confirm_if_needed(confirm: bool) -> Result<()> {
    if !confirm {
        return Ok(());
    }

    let proceed = Confirm::with_theme(&ColorfulTheme::default())
        .with_prompt("Do you really want to continue?")
        .default(true)
        .interact()
        .context("failed to ask confirmation")?;

    if proceed {
        return Ok(());
    }

    bail!("confirmation failed")
}

pub fn scrape_and_confirm(url: &str, confirm: bool) -> Result<String> {
    log::info!("Scraping data ...");
    let data = scrape(vec![url]).context("scrape failed")?;
    log::info!("Scraped data:\n{}", &data);

    confirm_if_needed(confirm)?;

    Ok(data)
}

#[cfg(test)]
mod tests {
    use anyhow::{Context, Result};
    use serde_json::Value;

    use super::scrape;

    fn scrape_and_parse(url: &str) -> Result<Value> {
        let result = scrape(vec![url, "production"])?;

        serde_json::from_str(&result).context("failed to parse")
    }

    #[test]
    #[ignore]
    fn test_scrape_book() -> Result<()> {
        let result =
            scrape_and_parse("https://www.yakaboo.ua/ua/stories-of-your-life-and-others.html")?;

        insta::assert_json_snapshot!("test_scrape_book", result);

        Ok(())
    }

    #[test]
    #[ignore]
    fn test_scrape_film() -> Result<()> {
        let result = scrape_and_parse("https://www.imdb.com/title/tt0133093/")?;

        insta::assert_json_snapshot!("test_scrape_film", result);

        Ok(())
    }
}
