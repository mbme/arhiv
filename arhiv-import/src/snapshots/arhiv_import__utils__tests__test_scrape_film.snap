---
source: arhiv-import/src/utils.rs
expression: result

---
{
  "cast": "Keanu Reeves, Laurence Fishburne, Carrie-Anne Moss",
  "countries_of_origin": "United States, Australia",
  "cover_src": "https://m.media-amazon.com/images/M/MV5BNzQzOTk3OTAtNDQ0Zi00ZTVkLWI0MTEtMDllZjNkYzNjNTc4L2ltYWdlXkEyXkFqcGdeQXVyNjU0OTQ0OTY@._V1_QL75_UX190_CR0,2,190,281_.jpg",
  "creators": "Lana Wachowski, Lilly Wachowski",
  "description": "Thomas A. Anderson is a man living two lives. By day he is an average computer programmer and by night a hacker known as Neo. Neo has always questioned his reality, but the truth is far beyond his imagination. Neo finds himself targeted by the police when he is contacted by Morpheus, a legendary computer hacker branded a terrorist by the government. As a rebel against the machines, Neo must confront the agents: super-powerful computer programs devoted to stopping Neo and the entire human rebellion.",
  "duration": "2h 16m",
  "is_series": "false",
  "original_language": "English",
  "release_date": "1999",
  "title": "The Matrix"
}
